feat: modify 1-bit data encoding

EscapedGibbon opened the pull request:
close: #20

EscapedGibbon committed:
feat: add encoder for 1 byte = 1 pixel data

EscapedGibbon committed:
test: modify test cases

EscapedGibbon committed:
test: add test to check encode/decode compatibility

EscapedGibbon committed:
refactor: modify test function name

EscapedGibbon commented:
There were some changes in the way the encoding was tested, because buffer from `fs.readFileSync()` was not giving the same result, when it came to partially filled data bytes. In my implementation the non-used bits were zeroed, while `readFileSync` version had random values. After discussion with @stropitek we arrived to the conclusion that those bits probably kept their previous value in memory. I haven't been able to find any citations about it, but it seems that there isn't any particular standard on how these bits must be encoded, since they have no effect on the encoded data. I did the bmp encoding in python, and its pixel data encoding was the same as mine(all 0s). Therefore, now the `testHeaderDataEncode` checks the data before pixel encoding, and pixel data is checked through snapshots.

EscapedGibbon commented:
I also looked at the images after my encoding and they seem to be the same. Since images are small and only have 2 colors, the differences would be noticeable visually if encoding did not work

stropitek commented:
If you regenerate an image with python, do you get exactly the same encoded bmp (header + pixels) as with this library?

In that case I would prefer that you regenerate all the images with python and not do snapshots.

EscapedGibbon commented:
> If you regenerate an image with python, do you get exactly the same encoded bmp (header + pixels) as with this library?
> 
> In that case I would prefer that you regenerate all the images with python and not do snapshots.

I don't get the same images with python, it has smaller header size. I think it might be because the python library that i used, uses v3 bitmap headers, when we use v5. I will look a bit more if there is a liabrary that supports v5 encoding

targos commented:
> readFileSync version had random values

This seems highly unlikely. I would be surprised if file systems store data in blocks that are not multiples of 8bits

stropitek commented:
> This seems highly unlikely. I would be surprised if file systems store data in blocks that are not multiples of 8bits

I think what he means is that when those bmp were generated (unfortunately I don't remember what program I used to generate them), they were initialized in memory with random bytes (not initialized to 0), so the bytes that don't encode anything were random when they were written to the disk.

EscapedGibbon commented:
> > This seems highly unlikely. I would be surprised if file systems store data in blocks that are not multiples of 8bits
> 
> I think what he means is that when those bmp were generated (unfortunately I don't remember what program I used to generate them), they were initialized in memory with random bytes (not initialized to 0), so the bytes that don't encode anything were random when they were written to the disk.

Yes, exactly

EscapedGibbon commented:
Gimp does export 1 bit images with v5 headers and its pixel data aligns with our implementation.  The header's data, however is slightly different. Judging by wikipedia specs and indexes of the bytes that were different I suspect that Gimp version also adds "horizonal and vertical pixel per metre" data.

![CleanShot 2025-03-21 at 12 01 16](https://github.com/user-attachments/assets/d30503a9-12e7-4bb8-8cb1-e5b2516f7f80)

stropitek commented:
Those resolution values seem useful, if we can we should decode and encode them.

targos commented:
I'd open an issue about this and specifically override these values before running the test assertion so this PR can be merged.

EscapedGibbon committed:
fix: align encoding and decoding with GIMP version

EscapedGibbon commented:
After discussion with @stropitek I regenerated images with GIMP. I also added the data on `pixelsPerMeter` and modified some header data that was not coherent with GIMP buffer. For now, changed values are not really important for 1 bit encoding/decoding so they were changed for alignment with GIMP version. Tests now pass by simply matching buffer from `fs.writeFileSync` with the buffer from our implementation. So snapshots are completely removed.

stropitek commented on src/BMPDecoder.ts:
```diff
@@ -21,6 +22,8 @@ export default class BMPDecoder {
     this.height = this.bufferData.readUint32();
     this.bitDepth = this.bufferData.seek(28).readUint16();
+    this.xPixelsPerM = this.bufferData.seek(38).readInt32();
+    this.yPixelsPerM = this.bufferData.readInt32();
```
Use full words: `xPixelsPerMeter`

stropitek commented on src/__test__/files/TestingGimp.bmp:
Is this file unused? Please remove all unused files.

EscapedGibbon committed:
docs: update comment

stropitek commented on src/__test__/bitDepth1.read.test.ts:
```diff
@@ -70,14 +79,15 @@ describe('decode image with bitDepth of 1', () => {
 
   it('decode a 62x4', () => {
     const result = decode(fs.readFileSync('src/__test__/files/62x4.bmp'));
-    expect(result).toMatchSnapshot();
+    testDecode(result, '62x4.bmp');
```
This test can be removed or changed to be like the others.

It's basically doing `expect(decode('62x4.bmp')).toEqual(decode('62x4.bmp'))`.

stropitek reviewed (commented):
Show us the commit message for this PR.

I would use `feat:` with the new resolution data which is encoded / decoded.

With the list of breaking changes regarding the codec format and the changes in the encoded data.

targos commented on src/BMPEncoder.ts:
```diff
@@ -27,19 +27,25 @@ export interface ImageCodec {
    * Image number of channels excluding alpha.
    */
   components: number;
+  /**
+   * Horizontal number of pixels per meter.
+   */
+  xPixelsPerMeter: number;
+  /**
+   * Vertical number of pixels per meter.
+   */
+  yPixelsPerMeter: number;
```
Shouldn't we have a default value? Seems wrong to make it mandatory.

EscapedGibbon replied:
If it's okay then i can make it 2835 as default value for both properties. It's the value for all tested images.

targos replied:
SGTM. This corresponds to 72dpi, which is indeed a legacy resolution.

EscapedGibbon committed:
fix: resolve conversations

EscapedGibbon commented:
> Show us the commit message for this PR.
> 
> I would use `feat:` with the new resolution data which is encoded / decoded.
> 
> With the list of breaking changes regarding the codec format and the changes in the encoded data.

Something like this. I am not sure if last two are a breaking change though. Added them just in case.

feat!: modify 1-bit image encoder

BREAKING CHANGES:

- In  `BMPEncoder`'s data each pixel is now stored as 1 byte per pixel instead of 1 bit per pixel.
- Optional fields `xPixelsPerMeter` and `yPixelsPerMeter` have been added to `ImageCodec`. Their default value is 2835.
-  `Decode()` function now also returns `xPixelsPerMeter` and `yPixelsPerMeter` in the `ImageCodec`.
- Class `BMPEncoder` doesn't inherit `IOBuffer` anymore.
- Class `BMPEncoder` now has `data` field to store Uint8Array data of the image.

stropitek commented:
You should write the commit message with the changelog it will generate in mind (this is what the users of the library will read if they upgrade).

What is in the main `feat:` message will be in the changelog under the "Features" category.

What is in `breaking-changes:` will be in the footer of the changelog.

My suggestion was to put everything related to the new encoded / decoded property in the `feat` main message, and also document the breaking changes related to the change of data structure in the breaking changes.

> * In  `BMPEncoder`'s data each pixel is now stored as 1 byte per pixel instead of 1 bit per pixel.

Isn't it more relevant to refer to the encoder's input and decoder's output? It's not obvious to me that the encoder's data is its input and not its output.

> * Optional fields `xPixelsPerMeter` and `yPixelsPerMeter` have been added to `ImageCodec`. Their default value is 2835

As @targos mentioned, it's a good idea to say it corresponds to 72 dpi which is a legacy resolution. Otherwise it just seems to be a random number. Also this is part of the added feature and is not a breaking change. Should be in the main commit message.

> * Class `BMPEncoder` now has `data` field to store Uint8Array data of the image.

Not a breaking change. But ok to mention it there if it is worth having it in the changelog.

targos commented:
>> Class BMPEncoder now has data field to store Uint8Array data of the image.

> Not a breaking change. But ok to mention it there if it is worth having it in the changelog.

The BMPEncoder class is not exposed publicly so this shouldn't be mentioned in the changelog.

EscapedGibbon commented:
feat: encode and decode resolution fields `xPixelsPerMeter` and `yPixelsPerMeter`

Encode / decode the image's horizontal and vertical number of pixels per meter. 

BREAKING CHANGE: In decoder's output data and encoder's input data, each pixel is now stored as 1 byte per pixel instead of 1 bit per pixel.

stropitek commented on src/constants.ts:
```diff
@@ -24,4 +24,6 @@ export const BITMAPV5HEADER = {
     LCS_GM_IMAGES: 0x00000004,
   },
+
+  DEFAULT_PIXELS_PER_METER: 2835,
```
Please add comment with how much that converts to in DPI

stropitek reviewed (approved):
You can merge when the last comment is resolved.

EscapedGibbon committed:
docs: add comment about default image resolution
